cdist/preos/debootstrap/files/devuan-debootstrap/debootstrap.8
.TH DEBOOTSTRAP 8 2015-05-21 "Devuan Project" "Devuan GNU/Linux manual"
.SH NAME
debootstrap \- Bootstrap a basic Devuan system
.SH SYNOPSIS
.B debootstrap
.RB [ OPTION\&.\&.\&. ]
.I SUITE TARGET
.RI [ MIRROR
.RI [ SCRIPT ]]

.B debootstrap
.RB [ OPTION\&.\&.\&. ]
\-\-second\-stage
.SH DESCRIPTION
.B debootstrap
bootstraps a basic Devuan system of
.I SUITE
into
.I TARGET
from
.I MIRROR
by running
.IR SCRIPT .
.I MIRROR
can be an http:// or https:// URL, a file:/// URL, or an ssh:/// URL.
.PP
The
.I SUITE
may be a release code name (eg, ceres, ascii, jessie)
or a symbolic name (eg, unstable, testing, stable, oldstable)
.PP
Notice that file:/ URLs are translated to file:/// (correct scheme as
described in RFC1738 for local filenames), and file:// will \fBnot\fR work.
ssh://USER@HOST/PATH URLs are retrieved using
.BR scp ;
use of
.B ssh\-agent
or similar is strongly recommended.
.PP
\fBDebootstrap\fR can be used to install Devuan in a system without using an
installation disk but can also be used to run a different Devuan flavor in a \fBchroot\fR
environment.
This way you can create a full (minimal) Devuan installation which
can be used for testing purposes (see the \fBEXAMPLES\fR section). 
If you are looking for a chroot system to build packages please take a look at 
\fBpbuilder\fR.
.SH "OPTIONS"
.PP
.IP "\fB\-\-arch=ARCH\fP"
Set the target architecture (use if dpkg isn't installed).
See also \-\-foreign.
.IP
.IP "\fB\-\-include=alpha,beta\fP"
Comma separated list of packages which will be added to download and extract
lists.
.IP
.IP "\fB\-\-exclude=alpha,beta\fP"
Comma separated list of packages which will be removed from download and
extract lists.
WARNING: you can and probably will exclude essential packages, be
careful using this option.
.IP
.IP "\fB\-\-components=alpha,beta\fP"
Use packages from the listed components of the archive.
.IP
.IP "\fB\-\-no\-resolve\-deps\fP"
By default, debootstrap will attempt to automatically resolve any missing
dependencies, warning if any are found.
Note that this is not a complete dependency resolve in the sense of dpkg
or apt, and that it is far better to specify the entire base system than
rely on this option.
With this option set, this behaviour is disabled.
.IP
.IP "\fB\-\-variant=minbase|buildd|fakechroot\fP"
Name of the bootstrap script variant to use.
Currently, the variants supported are minbase, which only includes
essential packages and apt; buildd, which installs the build-essential
packages into
.IR TARGET ;
and fakechroot, which installs the packages without root privileges.
The default, with no \fB\-\-variant=X\fP argument, is to create a base
Devuan installation in
.IR TARGET .
.IP
.IP "\fB\-\-merged-usr\fP"
Create /{bin,sbin,lib}/ symlinks pointing to their counterparts in /usr/.
.IP
.IP "\fB\-\-no-merged-usr\fP"
Do not create /{bin,sbin,lib}/ symlinks pointing to their counterparts in /usr/.
(Default.)
.IP
.IP "\fB\-\-keyring=KEYRING\fP"
Override the default keyring for the distribution being bootstrapped,
and use
.IR KEYRING
to check signatures of retrieved Release files.
.IP
.IP "\fB\-\-no-check-gpg\fP"
Disables checking gpg signatures of retrieved Release files.
.IP
.IP "\fB\-\-force-check-gpg\fP"
Forces checking Release file signatures, disabling automatic fallback to
HTTPS in case of a missing keyring. Incompatible with the previous option.
.IP
.IP "\fB\-\-verbose\fP"
Produce more info about downloading.
.IP
.IP "\fB\-\-print\-debs\fP"
Print the packages to be installed, and exit.
Note that a TARGET directory must be specified so debootstrap can
download Packages files to determine which packages should be installed,
and to resolve dependencies.
The TARGET directory will be deleted unless \-\-keep\-debootstrap\-dir
is specified.
.IP
.IP "\fB\-\-download\-only\fP"
Download packages, but don't perform installation.
.IP
.IP "\fB\-\-foreign\fP"
Do the initial unpack phase of bootstrapping only, for example if the
target architecture does not match the host architecture.
A copy of debootstrap sufficient for completing the bootstrap process
will be installed as /debootstrap/debootstrap in the target filesystem.
You can run it with the \fB\-\-second\-stage\fP option to complete the
bootstrapping process.
.IP
.IP "\fB\-\-second\-stage\fP"
Complete the bootstrapping process.
Other arguments are generally not needed.
.IP
.IP "\fB\-\-second\-stage\-target=DIR\fP"
Run second stage in a subdirectory instead of root. (can be used to create
a foreign chroot) (requires \-\-second\-stage)
.IP
.IP "\fB\-\-keep\-debootstrap\-dir\fP"
Don't delete the /debootstrap directory in the target after completing the
installation.
.IP
.IP "\fB\-\-unpack\-tarball=FILE\fP"
Acquire .debs from tarball FILE instead of downloading via http.
.IP
.IP "\fB\-\-make\-tarball=FILE\fP"
Instead of bootstrapping, make a tarball (written to FILE) of the downloaded
packages.
The resulting tarball may be passed to a later
.BR \-\-unpack\-tarball .
.IP
.IP "\fB\-\-debian\-installer\fP"
Used for internal purposes by the debian-installer
.IP 
.IP "\fB\-\-extractor=TYPE\fP"
Override automatic .deb extractor selection to
.IR TYPE .
Supported extractors are: dpkg-deb and ar.
.IP
.IP "\fB\-\-no\-check\-certificate\fP"
Do not check certificate against certificate authorities
.IP
.IP "\fB\-\-certificate=FILE\fP"
Use the client certificate stored in file (PEM)
.IP
.IP "\fB\-\-private\-key=FILE\fP"
Read the private key from file

.SH EXAMPLES
.
.PP 
To setup an \fIascii\fR system:
.PP 
     debootstrap ascii ./ascii-chroot http://packages.devuan.org/merged
.PP
     debootstrap ascii ./ascii-chroot file:///LOCAL_MIRROR/devuan
.PP
Full process to create a complete Devuan installation of \fIceres\fR (unstable)
in a chroot:
.PP
     main # debootstrap ceres ceres-root http://packages.devuan.org/merged
     [ ... watch it download the whole system ]
     main # echo "proc ceres-root/proc proc defaults 0 0" >> /etc/fstab
     main # mount proc ceres-root/proc -t proc
     main # echo "sysfs ceres-root/sys sysfs defaults 0 0" >> /etc/fstab
     main # mount sysfs ceres-root/sys -t sysfs
     main # cp /etc/hosts ceres-root/etc/hosts
     main # chroot ceres-root /bin/bash
.SH AUTHOR
.B debootstrap
This manpage was originally written by Anthony Towns <ajt@debian.org>.
and Matt Kraai <kraai@debian.org> for Debian.
It was re-worked for Devuan by Daniel Reurich <daniel@centurion.net.nz>.


doc/changelog
Changelog
---------

	* Changes are always commented with their author in (braces)
	* Exception: No braces means author == Nico Schottelius

2.0.15: 2012-11-02
	* Core: Make variable __object_name available in type explorers (Steven Armtrong)
	* New Type: __qemu_img
	* New Type: __line
	* New Type: __pf_apply (Jake Guffey)
	* New Type: __pf_ruleset (Jake Guffey)
	* Bugfix Type: __rvm: Make type work if rvm is already installed

2.0.14: 2012-09-07
	* Bugfix Type: __jail: Use correct variable (Jake Guffey)
	* Change Type: __jail: Parameter jailbase now optional (Jake Guffey)
	* Bugfix Type: __user: Use passwd database on FreeBSD (Jake Guffey)
	* Bugfix Type: __start_on_boot: Do not change parameters
	* Feature __user: Added support for BSDs (Sébastien Gross)
	* Feature __group: Added support for FreeBSD (Jake Guffey)
	* New Type: __package_zypper
	* Feature Types: Initial Support for SuSE Linux
	
2.0.13: 2012-06-05
	* Bugfix __ssh_authorized_key: Ensure it sets proper group (contradict)
	* Bugfix __addifnosuchline: Fixed quotes/interpolation bug ("a  b" became "a b")
	* New Explorer: interfaces (Sébastien Gross)
	* Feature core: Support reading from stdin in types (Steven Armstrong)
	* Feature core: Support multiple parameters for types (Steven Armstrong)
	* Feature __file: Support reading from stdin with - syntax (Steven Armstrong)

2.0.12: 2012-05-29
	* Core: Correctly raise error on Python < 3.2 (Steven Armtrong)
	* Core: Add support for --remote-exec and --remote-copy parameters
	* Documentation: Debian Squeeze hints (Sébastien Gross)

2.0.11: 2012-05-23
	* Fix insecure file/directory creation: Use umask 077

2.0.10: 2012-05-18
	* Cleanup __group: No getent gshadow in old Redhat, use groupmod -g
		(Matt Coddington)
	* Bugfix __package_yum: Missing cat
	* Bugfix __start_on_boot: Correctly use sed and quotes (Steven Armstrong)
	* Feature __file: Support for --state exists (Steven Armstrong)
	* Feature core: Make variable __manifest available to type manifests
	* Feature core: Correct parent dependency handling (Steven Armstrong)
	* Bugfix several types: Fix sed for FreeBSD (Istvan Beregszaszi)
	* New Type: __jail (Jake Guffey)
	* Change Type: __rvm*: --state present/absent not installed/remvoed (Evax Software)
	* Bugfix Type: __cron: Hide error output from crontab
	* Various smaller bugfixes (Chris Lamb)

2.0.9: 2012-03-12
	* Cleanup documentation: Fix environment variable list to be properly 
		displayed (Giel van Schijndel)
	* Cleanup documentation: Some minor corrections
	* New Type: __package_opkg (Giel van Schijndel)
	* New Type: __package_pkg_freebsd (Jake Guffey)
	* New Type: __mysql_database (Benedikt Koeppel)
	* Feature __package: Support for OpenWRT (Giel van Schijndel)
	* Feature __start_on_boot: Support for OpenWRT (Giel van Schijndel)
	* Feature __start_on_boot: Support for Amazon Linux (Matt Coddington)
	* New Example: Use rsync to backup files (Matt Coddington) 
	* Feature core: Exit non-zero, if configuration failed
	* Documentation: Describe how to do templating (Aurélien Bondis)

2.0.8: 2012-02-20
	* Bugfix core: Remove another nasty traceback when sending SIGINT (aka Ctrl-C)
	* Cleanup: Better hint to source of error
	* Cleanup: Do not output failing script, but path to script only
	* Cleanup: Remove support for __debug variable in manifests (Type != Core
		debugging)
	* Cleanup: Change __package_* to support absent/present (default state
		name now). The values removed/installed will be removed in cdist 2.1.
	* Cleanup: Change __process to support absent/present (default state
		name now). The values running/stopped will be removed in cdist 2.1.
	* Feature Core: Support boolean parameters (Steven Armstrong)

2.0.7: 2012-02-13
	* Bugfix __file: Use chmod after chown/chgrp (Matt Coddington)
	* Bugfix __user: Correct shadow field in explorer (Matt Coddington)
	* Bugfix __link: Properly handle existing links (Steven Armstrong)
	* Bugfix __key_value: More robust implementation (Steven Armstrong)
	* Bugfix __user: Fix for changing a user's group by name (Matt Coddington)
	* New Type: __package_pip
	* Bugfix/Cleanup: Correctly allow Object ID to start and end with /, but
		not contain //.

2.0.6: 2012-01-28
	* Bugfix __apt_ppa:
		Also remove the [ppa-name].list file, if empty. (Tim Kersten)
	* Bugfix __group:
		Referenced wrong variable name (Matt Coddington)
	* Feature __package_apt:
		Initial support for virtual packages (Evax Software)
	* Feature Core: Added new dependency resolver (Steven Armstrong)
	* Feature Explorer, __package_yum: Support Amazon Linux (Matt Coddington)
	* New Type: __rvm (Evax Software)
	* New Type: __rvm_gem (Evax Software)
	* New Type: __rvm_gemset (Evax Software)
	* New Type: __rvm_ruby (Evax Software)

2.0.5: 2012-01-18
	* Bugfix __key_value: Use correct delimiters
		(Steven Armstrong, Daniel Maher)
	* Cleanup: Explicitly require Python >= 3.2 (do not fail implicitly)
	* Documentation: (Re)write of the tutorial
	* Feature: __addifnosuchline supports matching on 
		regular expressions (Daniel Maher)
	* Feature: __directory, __file, __link:
		Add --state parameter (Steven Armstrong)
	* New Type: __package_luarocks (Christian G. Warden)
	* New Type: __cdistmarker (Daniel Maher)

2.0.4: 2011-11-18
	* Bugfix core: Remove traceback when sending SIGINT (aka Ctrl-C)
	* Bugfix core: Accept parameters with - in the name (Steven Armstrong)
	* Cleanup: __object_fq variable removed (never used)
	* Cleanup: Environment variable __self DEPRECATED, use __object_name instead
	* Cleanup: Environment variable __self scheduled for removal in cdist 2.1
	* Documentation: Many examples for use of __remote_* (Steven Armstrong)
	* Feature: Automatically require all used objects (Steven Armstrong)
	* New Type: __cron (Steven Armstrong)

2.0.3: 2011-10-18
	* Improved logging, added --verbose, by more quiet by default
	* Bugfix __user: Correct quoting (Steven Armstrong)
	* Bugfix __addifnosuchline: Falsely exited

cdist/preos/debootstrap/debootstrap.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# 2016 Darko Poljak (darko.poljak at ungleich.ch)
#
# This file is part of cdist.
#
# cdist is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# cdist is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with cdist. If not, see <http://www.gnu.org/licenses/>.
#
#

import cdist
import cdist.config
import cdist.core
import cdist.preos
import argparse
import cdist.argparse
import logging
import os
import subprocess


class Debian(object):
    _preos_name = 'debian'
    _cdist_preos = True

    _files_dir = os.path.join(os.path.dirname(__file__), "files")

    @classmethod
    def default_args(cls):
        default_remote_exec = os.path.join(cls._files_dir, "remote-exec.sh")
        default_remote_copy = os.path.join(cls._files_dir, "remote-copy.sh")
        default_init_manifest = os.path.join(
            cls._files_dir, "init-manifest-{}".format(cls._preos_name))

        defargs = argparse.Namespace()
        defargs.arch = 'amd64'
        defargs.bootstrap = False
        defargs.configure = False
        defargs.cdist_params = '-v'
        defargs.rm_bootstrap_dir = False
        defargs.suite = 'stable'
        defargs.remote_exec = default_remote_exec
        defargs.remote_copy = default_remote_copy
        defargs.manifest = default_init_manifest

        return defargs

    @classmethod
    def get_parser(cls):
        defargs = cls.default_args()
        cdist_parser = cdist.argparse.get_parsers()
        parser = argparse.ArgumentParser(
                prog='cdist preos {}'.format(cls._preos_name),
                parents=[cdist_parser['loglevel'], cdist_parser['beta']])
        parser.add_argument('target_dir', nargs=1,
                            help=("target directory where PreOS will be "
                                  "bootstrapped"))
        parser.add_argument(
            '-a', '--arch',
            help="target debootstrap architecture, by default '{}'".format(
                defargs.arch), dest='arch', default=defargs.arch)
        parser.add_argument(
            '-B', '--bootstrap',
            help='do bootstrap step',
            dest='bootstrap', action='store_true', default=defargs.bootstrap)
        parser.add_argument(
            '-C', '--configure',
            help='do configure step',
            dest='configure', action='store_true', default=defargs.configure)
        parser.add_argument(
            '-c', '--cdist-params',
            help=("parameters that will be passed to cdist config, by default"
                  " '{}' is used".format(defargs.cdist_params)),
            dest='cdist_params', default=defargs.cdist_params)
        parser.add_argument(
            '-D', '--drive-boot',
            help='create bootable PreOS on specified drive',
            dest='drive')
        parser.add_argument(
            '-e', '--remote-exec',
            help=("remote exec that cdist config will use, by default "
                  "internal script is used"),
            dest='remote_exec', default=defargs.remote_exec)
        parser.add_argument(
            '-i', '--init-manifest',
            help=("init manifest that cdist config will use, by default "
                  "internal init manifest is used"),
            dest='manifest', default=defargs.manifest)
        parser.add_argument(
            '-k', '--keyfile', action="append",
            help=("ssh key files that will be added to cdist config; "
                  "'__ssh_authorized_keys root ...' type is appended to "
                  "initial manifest"),
            dest='keyfile')
        parser.add_argument(
            '-m', '--mirror',
            help='use specified mirror for debootstrap',
            dest='mirror')
        parser.add_argument(
            '-P', '--root-password',
            help='Set specified password for root, generated by default',
            dest='root_password')
        parser.add_argument('-p', '--pxe-boot-dir', help='PXE boot directory',
                            dest='pxe_boot_dir')
        parser.add_argument(
            '-r', '--rm-bootstrap-dir',
            help='remove target directory after finishing',
            dest='rm_bootstrap_dir', action='store_true',
            default=defargs.rm_bootstrap_dir)
        parser.add_argument(
            '-S', '--script',
            help='use specified script for debootstrap',
            dest='script')
        parser.add_argument('-s', '--suite',
                            help="suite used for debootstrap, "
                                 "by default '{}'".format(defargs.suite),
                            dest='suite', default=defargs.suite)
        parser.add_argument(
            '-t', '--trigger-command',
            help=("trigger command that will be added to cdist config; "
                  "'__cdist_preos_trigger http ...' type is appended to "
                  "initial manifest"),
            dest='trigger_command')
        parser.add_argument(
            '-y', '--remote-copy',
            help=("remote copy that cdist config will use, by default "
                  "internal script is used"),
            dest='remote_copy', default=defargs.remote_copy)
        parser.epilog = cdist.argparse.EPILOG

        return parser

    @classmethod
    def update_env(cls, env):
        pass

    @classmethod
    def commandline(cls, argv):
        log = logging.getLogger(cls.__name__)

        parser = cls.get_parser()
        args = parser.parse_args(argv)
        if args.script and not args.mirror:
            raise cdist.Error("script option cannot be used without "
                              "mirror option")

        args.command = cls._preos_name
        cdist.argparse.check_beta(vars(args))

        cdist.preos.check_root()

        args.target_dir = os.path.realpath(args.target_dir[0])
        args.os = cls._preos_name
        args.remote_exec = os.path.realpath(args.remote_exec)
        args.remote_copy = os.path.realpath(args.remote_copy)
        args.manifest = os.path.realpath(args.manifest)
        if args.keyfile:
            new_keyfile = [os.path.realpath(x) for x in args.keyfile]
            args.keyfile = new_keyfile
        if args.pxe_boot_dir:
            args.pxe_boot_dir = os.path.realpath(args.pxe_boot_dir)

        cdist.argparse.handle_loglevel(args)
        log.debug("preos: {}, args: {}".format(cls._preos_name, args))
        try:
            env = vars(args)
            new_env = {}
            for key in env:
                if key == 'verbose' and env[key]:
                    if env[key] >= 3:
                        new_env['debug'] = "yes"
                    elif env[key] == 2:
                        new_env['verbose'] = "yes"
                elif not env[key]:
                    new_env[key] = ''
                elif isinstance(env[key], bool) and env[key]:
                    new_env[key] = "yes"
                elif isinstance(env[key], list):
                    val = env[key]
                    new_env[key + "_cnt"] = str(len(val))
                    for i, v in enumerate(val):
                        new_env[key + "_" + str(i)] = v
                else:
                    new_env[key] = str(env[key])
            env = new_env
            env.update(os.environ)
            cls.update_env(env)
            log.debug("preos: {} env: {}".format(cls._preos_name, env))
            cmd = os.path.join(cls._files_dir, "code")
            info_msg = ["Running preos: {}, suite: {}, arch: {}".format(
                cls._preos_name, args.suite, args.arch), ]
            if args.mirror:
                info_msg.append("mirror: {}".format(args.mirror))
            if args.script:
                info_msg.append("script: {}".format(args.script))
            if args.bootstrap:
                info_msg.append("bootstrapping")
            if args.configure:
                info_msg.append("configuring")
            if args.pxe_boot_dir:
                info_msg.append("creating PXE")
            if args.drive:
                info_msg.append("creating bootable drive")
            log.info(info_msg)
            log.debug("cmd={}".format(cmd))
            subprocess.check_call(cmd, env=env, shell=True)
        except subprocess.CalledProcessError as e:
            log.error("preos {} failed: {}".format(cls._preos_name, e))


class Ubuntu(Debian):
    _preos_name = "ubuntu"

    @classmethod
    def default_args(cls):
        defargs = super().default_args()
        defargs.suite = 'xenial'
        return defargs


class Devuan(Debian):
    _preos_name = "devuan"

    @classmethod
    def default_args(cls):
        defargs = super().default_args()
        defargs.suite = 'jessie'
        return defargs

    @classmethod
    def update_env(cls, env):
        env['DEBOOTSTRAP_DIR'] = os.path.join(cls._files_dir,
                                              'devuan-debootstrap')
